Fix tab position not honored when set on add.

Kershoc opened the pull request:
<!-- Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows with the necessary information: -->

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | When adding a new tab any position set is not honored, resulting tab being added with lowest position.
| Type?         | bug fix 
| Category?     | CO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | 
| How to test?  | Have a module create a new tab with a position preset.  Before fix, the new tab will always be on bottom.  After fix, it will show where you expect.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/8431)
<!-- Reviewable:end -->

Kershoc committed:
CO: Fix tab position not honored when set on add.

Kershoc commented:
No worries Codacy-Bot.  If you want some space, I'll give you some space.

Quetzacoalt91 commented:
Hello @Kershoc,

I understand what you mean by this PR, but I'm not sure this is the best way to handle that.
An issue I can guess is two entries with the same position. We do not know how the shop handle this. 

A tab may disappear, or they will be randomly displayed?
We do not really know, but we can agree that having two tabs at the same position is ambiguous. I would prefer another option, where we can ask our tab to be set at the end or the beginning of the list. It could use two different method of placement, something like `STORE_AS_FIRST` and `STORE_AS_LAST`. If we store it at the top of the list, the core would automatically update the values of the other tabs.

Do you have other use cases to cover?

Kershoc commented:
The use case that originated this PR was that I was trying to add a Tab after another tab as it was related and logically looked better being under it in the menu.  That is how I noticed that the add method would overwrite the position to always be on bottom.

I get what you are saying about the ambiguity of having non unique values.  Especially if you are using that column for more then just order by.  With that in mind, there may actually be an issue with the update method as well.  

You can change the position to any value you wish on an existing tab and save it.  The update method will happily save whatever position value was passed to it.  It looks like update only overrides the position to a new value if the tab is moving from one parent to another.

Based on this new information an idea hit me that would work for first, last, and somewhere in between.  I'll work on that and update this PR.

kpodemski commented:
IMO this issue is more complex, we would need to check every element < position and > position and update it. 

I think that for such a cases the best way to handle it is to prevent position to have +1 but for example +10, having it this way we can manage position between two items, with that we can have something like:

$tab->position = $tab->insertBefore('AdminSomeTab'); // method get position of AdminSomeTab and add +1

Kershoc commented:
@kpodemski Exactly what I was thinking as well.  

Where my thoughts were going would be to check to see if we requested a specific position during add and if so, save the tab as normal at the bottom.  Then call updatePosition to put it where it was requested and reorder the list.  

Then something similar for update, check if the requested position was different from the current position and if so, keep it in its current position and once saved call updatePosition to move it to the requested spot.

Top it all off with a helper method(s) to get a +/- 1 of a current tabs position with out of bounds checks so we can do as you described.

This is all pseudo in my head my head right now, but it seemed like it would cover the bases.  I am more then open to any other avenues or ideas.

kpodemski commented:
@Kershoc i'm 100% with your idea but unless you'll create some PR with that solution i don't think that anyone from PrestaShop team will have time to create something

Kershoc committed:
Allow custom positions on add/update

comes with set helper methods to help quickly determine a position among other tabs

Kershoc commented:
Okay I have had some time to be able to get the pseudo out of my head and into actual code.  The updated commit to this PR will allow custom positions when adding and/or updating a tab, while keeping position number unique.  It uses the already existing method updatePosition() to set the tab into the requested position.  It covers the possible use cases first, last, and somewhere in between.

Position can now be specified on a new tab and will be honored by the add() method.  A position can be changed on an existing tab and will properly keep the positions unique during update().

Three new helper methods are added for developers, so they can quickly determine a position they want among other tabs.  getPositionByClassName() getPositionBeforeClassName() getPositionAfterClassName()

LittleBigDev commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[classes/Tab.php, line 94 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8431#-Kz2zSqly9g9dAsBxsHD:-Kz2zSqly9g9dAsBxsHE:b-po48rl) ([raw file](https://github.com/prestashop/prestashop/blob/2e54535755e7c27f007dcccf18449a86d95b7b06/classes/Tab.php#L94)):*
> ```PHP
> 
>         // Set good position for new tab
>         (is_int($this->position) && $this->position > 0) ? $requestedPosition = $this->position : $requestedPosition = null;
> ```

For readability purpose, I'd rather init `$requestedPosition = null`, and then update it if `$this->position` is positive.

---

*[classes/Tab.php, line 622 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8431#-Kz35hoy_Jl4_8onQub5:-Kz35hoz4S3MiI1OnD0C:bh3670k) ([raw file](https://github.com/prestashop/prestashop/blob/2e54535755e7c27f007dcccf18449a86d95b7b06/classes/Tab.php#L622)):*
> ```PHP
>         $res = parent::update($nullValues);
> 
>         if ($res && is_int($requestedPosition) && $requestedPosition > 0 && $requestedPosition < Tab::getNewLastPosition($this->id_parent)) {
> ```

If statement too long.
Please split this statement into multiple lines. We usually use this arrangement :
```php
if ($condition
    && $condition2
    || $condition3
) {
    // Code.
}
```
Another/additional way is to extract a private method dedicated to testing :
```php
if ($res && $this->shouldUpdatePosition($requestedPosition)) {
    // Code.
}
```

---

*[classes/Tab.php, line 726 at r2](https://reviewable.io:443/reviews/prestashop/prestashop/8431#-Kz3AegO-9wDEmGlzOs_:-Kz3AegO-9wDEmGlzOsa:b9cs7hg) ([raw file](https://github.com/prestashop/prestashop/blob/2e54535755e7c27f007dcccf18449a86d95b7b06/classes/Tab.php#L726)):*
> ```PHP
>             $newPosition = (int)$position - 1;
>             if ($newPosition <= 0) {
>                 $newPosition = 1;
> ```

What if position 1 is already assigned ?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8431)*
<!-- Sent from Reviewable.io -->

LittleBigDev reviewed (changes requested):
Thank you @Kershoc, and well done !
Only few maintainability suggestions and we are done.

Kershoc committed:
Implement Requested Maintainability Suggestions

Kershoc commented:
I have implemented the requested changes.  

As for the last question.  The positions returned by the new helper methods will in most cases be positions already in use.  updatePosition() takes care of reordering during save.  The helper methods are there so module developers can do something like:

```
$tab = new Tab();
$tab->position = Tab::getPositionAfterClassName("SomeClassName");
```
Without having to do their own calculations for desired position.

LittleBigDev commented:
Ok got your point.

LittleBigDev commented:
Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[classes/Tab.php, line 630 at r3](https://reviewable.io:443/reviews/prestashop/prestashop/8431#-Kz888Bh4E2KkLLL3FkW:-Kz888Bi3ohqaHXaoiCm:b-980oxh) ([raw file](https://github.com/prestashop/prestashop/blob/c7b60a771c219ea22430f81d2586e63dc6f45935/classes/Tab.php#L630)):*
> ```PHP
>             && $requestedPosition < Tab::getNewLastPosition($this->id_parent)
>         ) {
>             ($requestedPosition < $this->position) ? $way = false : $way = true;
> ```

Sorry I missed that !
```php
$way = $requestedPosition >= $this->position;
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8431)*
<!-- Sent from Reviewable.io -->

LittleBigDev reviewed (changes requested):
Really sorry I missed that :S

Kershoc committed:
Implement Requested Maintainability Suggestion

Kershoc commented:
No worries.  I am more then happy to make as many changes as needed.  😃

LittleBigDev commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8431#-:-KzNi2NlnLWzemsB0uAZ:bnfp4nl)*
<!-- Sent from Reviewable.io -->

LittleBigDev reviewed (dismissed):
Good Job !

Quetzacoalt91 commented:
The feature itself is validated, perfect.

However, we can't merge the PR without some tests. And I have to look for the automatic tab installation, I wonder how I could link your feature to it (https://github.com/PrestaShop/PrestaShop/pull/5922)

Quetzacoalt91 commented on classes/Tab.php:
```diff
@@ -684,4 +707,50 @@ public static function getTabModulesList($idTab)
+     * @param $className
+     * @return bool|int
+     */
+    public static function getPositionBeforeClassName($className)
```
I'm wondering ... If you want to set a tab before another one, don't you need to set the same position instead?

Kershoc replied:
You're right.  Since the tabs will be shifting to make room 'before' should be the same position.  I have corrected the method.

Quetzacoalt91 commented:
@Kershoc, I made some tests for your new functions `Tab::getPositionBeforeClassName()` and `Tab::getPositionAfterClassName()`, and maybe I missed something. Please check my other comment.

The tests I made are available here: https://github.com/Quetzacoalt91/PrestaShop/commit/797a85070e444bb4eff4baa43696d14864880284, and they are currently in failure for the following reasons:
```
There were 2 failures:

1) Tests\Unit\Classes\TabTest::testPositionBeforeClass with data set #0 (0, 'AdminModules')
Expected Tab::getPositionBeforeClassName(AdminModules) to return 0, got 1 instead.
Failed asserting that 1 is identical to 0.

/home/thomas/Documents/github/PrestaShop/tests/Unit/Classes/TabTest.php:68

2) Tests\Unit\Classes\TabTest::testPositionBeforeClass with data set #1 (0, 'AdminOrders')
Expected Tab::getPositionBeforeClassName(AdminOrders) to return 0, got 1 instead.
Failed asserting that 1 is identical to 0.

/home/thomas/Documents/github/PrestaShop/tests/Unit/Classes/TabTest.php:68
```

Did you know 0 is a valid position?
![capture du 2017-11-20 11-11-42](https://user-images.githubusercontent.com/6768917/33016951-5e691eca-cde8-11e7-9d33-8d60182b1fa3.png)

Kershoc committed:
minor corrections

fixes 0 not treated as valid position
fixes getPositionBeforeClassName() off by 1 error. If you want to insert into the position before a tab, you technically want that tabs current position not the one before it.

Kershoc commented:
I did not know 0 was valid.  I had wrongly assumed 1-n after noticing the cleanPositions method.  Entirely my fault for not checking against the database.  I have modified my work to treat 0 as a valid position.

I have also changed getPositionBeforeClassName method to return the requested tabs position.  Since the tabs will be shifting to make room 'before' should be the same position as the requested tab.

LittleBigDev commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> again.

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/prestashop/prestashop/8431#-:-KzbruGuScRf-sjHAvWt:butqosd)*
<!-- Sent from Reviewable.io -->

marionf commented:
Hello @Kershoc 
Can you provide a module, allowing to test your PR ?
Thanks :)

PierreRambaud commented:
Hi @Kershoc ,

Since we had no news from you for more than 30 days, I'll close this pull request. Feel free to reopen or open another one if you think it's still relevant.

Thanks!
